src/libstd/time/iso_8601.rs

```diff
@@ -0,0 +1,300 @@
+        debug_assert!((year % 400 + 400) % 400 == 370);
+
+        // Subtract 370 years to get to cycle start
+        // and additional 400 year to prevent integer underflow in duration subtraction
+        let cycle_start = year - 400 - 370;
+        let add_from_cycle_start =
+            Duration::from_secs(CYCLE_SECONDS + YEARS_1600_1970_SECONDS) - sub;
+
+        debug_assert!(cycle_start % 400 == 0);
+
+        return TmUtc::from_cycle_start_add_duration(cycle_start, add_from_cycle_start);
+    }
+}
+
+impl fmt::Display for TmUtc {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}",
+            self.year, self.month, self.day, self.hour, self.minute, self.second)?;
+
+        // if width is not specified, print nanoseconds
+        let subsec_digits = f.precision().unwrap_or(9);
+        if subsec_digits != 0 {
+            let mut subsec_digits = subsec_digits;
+
+            let width = if subsec_digits > 9 { 9 } else { subsec_digits };
+
+            // "Truncated" nanonseconds.
+            let mut subsec = self.nanos;
+
+            // Performs 8 iterations when width=1,
+            // but that's probably not a issue compared to other computations.
+            for _ in width..9 {
+                subsec /= 10;
+            }
+
+            write!(f, ".{:0width$}", subsec, width=width as usize)?;
+
+            // Adding more than 9 digits is meaningless,
+            // but if user requests it, we should print zeros.
+            for _ in 9..subsec_digits {
+                write!(f, "0")?;
+                subsec_digits -= 1;
+            }
+        }
+
+        write!(f, "Z")
+    }
+}
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use time::Duration;
+    use u64;
+
+    #[test]
+    fn test_fmt() {
+        fn test_impl(expected: &str, secs: i64, nanos: i32, subsec_digits: u32) {
+            if nanos != 0 {
+                assert_eq!(secs >= 0, nanos >= 0);
+            }
+
+            let time = if secs >= 0 {
+                TmUtc::from_epoch_add(Duration::new(secs as u64, nanos as u32))
+            } else {
+                TmUtc::from_epoch_sub(Duration::new(-secs as u64, -nanos as u32))
+            };
+
+            assert_eq!(expected, format!("{:.digits$}", time, digits=subsec_digits as usize));
+        }
+
+        // Tests can be validated with with GNU date:
+        // `TZ=UTC gdate --date @1535585179 --iso-8601=seconds`
+
+        test_impl("1970-01-01T00:00:00Z", 0, 0, 0);
+        test_impl("2018-08-29T23:26:19Z", 1535585179, 0, 0);
+        test_impl("2018-08-29T23:26:19.123Z", 1535585179, 123456789, 3);
+        test_impl("1970-01-01T00:00:00.000000001000Z", 0, 1, 12);
+        test_impl("5138-11-16T09:46:40Z", 100000000000, 0, 0);
+        test_impl("33658-09-27T01:46:41Z", 1000000000001, 0, 0);
+        // Largest value GNU date can handle
+        test_impl("2147485547-12-31T23:59:59Z", 67768036191676799, 0, 0);
```

Comment:
According to wikipedia, years more than 4 digits should have a leading `+`: https://en.wikipedia.org/wiki/ISO_8601#Years.  I haven't actually read the standard so I don't know if this is required or recommended or what.

Reply:
I think you are right. I've looked in the standard, and it uses a sign everywhere when a year has more than 4 digits. Also, Java implementation adds a sign: https://repl.it/repls/SphericalPrevailingTrapezoids Thank you!